hs_access_control/models.py

```diff
@@ -1685,81 +1510,68 @@ def share_resource_with_user(self, this_resource, this_user, this_privilege):
+        # cannot downgrade privilege just by having sharing privilege. 
+      
+        grantor_priv = access_resource.get_effective_privilege(self.user)
         grantee_priv = access_resource.get_effective_privilege(this_user)
 
         # check for user authorization
         if self.user.is_superuser:
             pass  # admin can do anything
 
-        elif whom_priv == PrivilegeCodes.OWNER:
+        elif grantor_priv == PrivilegeCodes.OWNER:
             pass  # owner can do anything
 
         elif access_resource.shareable:
-            pass  # non-owners can share
-
-        else:
-            raise PermissionDenied("User must own resource or have sharing privilege")
+            if grantor_priv > PrivilegeCodes.VIEW:
+                raise PermissionDenied("User has no privilege over resource")
 
-        # privilege checking
+            if grantor_priv > this_privilege:
+                raise PermissionDenied("User has insufficient privilege over resource")
 
-        if whom_priv > PrivilegeCodes.VIEW:
-            raise PermissionDenied("User has no privilege over resource")
+            # enforce non-idempotence for unprivileged users 
+            if grantee_priv == this_privilege: 
+                raise PermissionDenied("Non-owners cannot reshare at existing privilege")
 
-        if whom_priv > this_privilege:
-            raise PermissionDenied("User has insufficient privilege over resource")
+            if this_privilege > grantee_priv and this_user != self.user: 
+                raise PermissionDenied("Non-owners cannot decrease privileges for others")
 
-        # non owner can't downgrade privilege granted by someone else
-        # grantee_priv: current privilege of the grantee (this_user)
-        # this_privilege: proposed privilege for the grantee (this_user)
-        if whom_priv != PrivilegeCodes.OWNER and grantee_priv < this_privilege:
-            record = UserResourcePrivilege.objects.get(resource=this_resource,
-                                                       user=this_user,
-                                                       privilege=grantee_priv)
+        else:
+            raise PermissionDenied("User must own resource or have sharing privilege")
 
-            # current grantor (self) is not the same as the original grantor
-            if record.grantor != self.user:
-                raise PermissionDenied("User has insufficient privilege over resource")
+        # regardless of privilege, cannot remove last owner 
+        if grantee_priv == PrivilegeCodes.OWNER \
+                and this_privilege != PrivilegeCodes.OWNER \
+                and access_resource.owners.count() == 1: 
+            raise PermissionDenied("Cannot remove sole owner of resource")
 
-        # This logic implicitly limits one to one record per resource and requester.
-        with transaction.atomic():
+        # This logic implicitly limits one to one record per resource and grantee.
```

Comment:
Again it would be better to replace all the code above that do the checks with a call to ```can_share_resource_with_user()```